### DISCUSSION TITLE
How are missing bounds handled in xarray and xcdat ?

### DISCUSSION THREAD

**Discussion by jypeter** [xCDAT/xcdat#411] [2. Questions]
My apologies in this topic is already covered in xarray and xcdat documentation, just point me in the right direction! I confess I'm still not using xarray/xcdat and have only looked at very short sections of their documentation

Anyway, @mfwehner [comment](https://github.com/xCDAT/xcdat/discussions/404#discussioncomment-4770047) triggered some bounds' related questions. **How are missing bounds handled by xarray and xcdat** ?

In a perfect world, the netCDF files have perfect metadata and everything is handled as it should be, by default. In the real world, I indeed had to regularly use [cdms2.setAutoBounds](https://cdms.readthedocs.io/en/latest/generated/cdms2.axis.setAutoBounds.html) and `cdutil.setTimeBoundsXXXX`

I have failed to find some online links to `cdutil.setTimeBoundsXXXX` documentation, so the best way is to have a look at the [Old Documents](https://github.com/xCDAT/xcdat/discussions/170) I have put online (e.g. page 27 of `CDAT_Refresher_2010.pdf`, but there several other interesting references to _bounds_ in this pdf), or at the docstrings
```

>>> import cdutil

>>> dir(cdutil)
['ANNUALCYCLE', 'APR', 'ASeason', 'AUG', 'AveragerError', 'DEC', 'DJF', 'FEB', 'JAN', 'JJA', 'JUL', 'JUN', 'MAM', 'MAR', 'MAY', 'MV2', 'NOV', 'OCT', 'SEASONALCYCLE', 'SEP', 'SON', 'Seasons', 'TimeSlicer', 'ValidationFunctions', 'VariableConditioner', 'VariablesMatcher', 'WeightedGridMaker', 'WeightsMaker', 'YEAR', '__builtins__', '__cached__', '__check_weightoptions', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'area_weights', 'averager', 'basestring', 'cdat_info', 'cdms2', 'cdtime', 'centroid', 'create_landsea_mask', 'cyclicalcentroid', 'generalCriteria', 'generateLandSeaMask', 'generateSurfaceTypeByRegionMask', 'getAxisWeight', 'getAxisWeightByName', 'getMonthIndex', 'getMonthString', 'insert_monthly_seasons', 'isMonthly', 'linearInterpolation', 'logLinearInterpolation', 'mergeTime', 'monthBasedSlicer', 'numpy', 'reconstructPressureFromHybrid', 'region', 'retrieve', 'setAxisTimeBoundsDaily', 'setAxisTimeBoundsMonthly', 'setAxisTimeBoundsYearly', 'setSlabTimeBoundsDaily', 'setSlabTimeBoundsMonthly', 'setSlabTimeBoundsYearly', 'setTimeBoundsDaily', 'setTimeBoundsMonthly', 'setTimeBoundsYearly', 'sftbyrgn', 'sigma2Pressure', 'switchCalendars', 'times', 'vertical', 'weekday']

>>> help(cdutil.setTimeBoundsMonthly)
Help on function setTimeBoundsMonthly in module cdutil.times:

setTimeBoundsMonthly(obj, stored=0)
    Sets the bounds correctly for the time axis (beginning to end of month)
    Set stored to 1 to indicate that your data are stored at the end of the month
    Usage:
    tim=s.getTime()
    cdutil.times.setAxisTimeBoundsMonthly(s,stored=0)
    or
    cdutil.times.setAxisTimeBoundsMonthly(tim,stored=0)
```

Googling `xarray dimensions bounds` helped me find
- [xarray.Dataset.cf.add_bounds](https://cf-xarray.readthedocs.io/en/latest/generated/xarray.Dataset.cf.add_bounds.html). I would not rely too much on _The bounds values are guessed assuming equal spacing on either side of a coordinate label_...
- [xarray.Dataset.bounds.add_missing_bounds](https://xcdat.readthedocs.io/en/latest/generated/xarray.Dataset.bounds.get_bounds.html) in xcdat documentation. But I'm not too sure what you would get with a width different from the default (_Width of the bounds relative to the position of the nearest points, by default 0.5_). And you are likely to get wrong bounds for `T` or `Z` (well there is not much that can be done for height/depth anyway), even if it's better than nothing (or is it? maybe missing bounds have to trigger a warning)

**Comment by pochedls** [comment]
You have the option of adding bounds using [`xcdat.open_dataset()`](https://xcdat.readthedocs.io/en/latest/generated/xcdat.open_dataset.html) (by default missing bounds are added) or they can be explicitly added using [`ds.bounds.add_bounds()`](https://xcdat.readthedocs.io/en/latest/generated/xarray.Dataset.bounds.add_bounds.html). 

I think the `add_bounds` functionality was refactored from CDAT into xCDAT (CDAT assumed a "width" of 0.5, which puts the bounds halfway between the axis coordinates). 

I don't think we have anything equivalent to `...setAxisTimeBoundsXYZ` – I think this would be a good feature request.

**Reply by pochedls** [reply]
In thinking about this discussion thread and @mfwehner's comment, I think xcdat should probably limit the current `add_missing_bounds` routine to latitude and longitude (this is currently applied to all axes that are missing bounds if `add_bounds=True` in `xcdat.open_dataset`) and implement some version of CDAT's `setTimeBoundsFREQ`. I played with CDAT v8.2 today and it does not automatically detect the appropriate frequency of missing time bounds (the user has to decide between `setAxisTimeBoundsDaily`, `setAxisTimeBoundsMonthly`, `setAxisTimeBoundsYearly`) – so the user needs to call one of these methods as appropriate. I'll add a feature request for this, but please add to the discussion if you have further comments on this.

**Reply by jypeter** [reply]
You could control the bounds' generation behavior with something similar to [cdms2.setAutoBounds](https://cdms.readthedocs.io/en/latest/generated/cdms2.axis.setAutoBounds.html), and print some warnings when routines depending on correct bounds find out that no bounds are available or that the available bounds were automatically generated.

I have just checked that the default behavior of `AutoBounds` is indeed `2 : ‘grid’ (Generate bounds for lat/lon grids only)`
```
>>> import cdms2
>>> cdms2.getAutoBounds()
2
```

**Reply by pochedls** [reply]
I've tried to capture this thread in Feature Request #412.

**Comment by mfwehner** [comment]
It strikes me as hard to automatically detect the bounds except in simple cases. If units are "months since xx" and the increment is 1, then clearly monthly or if "days since xx” and unit increment then clearly days. But if monthly data and the units are “day since xx” then you would probably have to crack the calendar etc.

What I would prefer is that you catch the problem and print out a clear error message. Like you need to run setAxisTimeBoundsXX etc.

M

**Reply by pochedls** [reply]
What made me wonder about whether this could be detected was an [`isMonthly`](https://github.com/CDAT/cdutil/blob/master/cdutil/times.py#L166) function I found in cdutil, but I think I agree with you that the user should get a warning and call the appropriate `setAxisTimeBoundsXYZ` function on their own.

**Comment by jypeter** [comment]
I have asked @oliviermarti and he thinks automatically generating bounds for `Z` vertical levels (depth or height) can probably also be a source of errors and should not be done by default
